packages/asset-swapper/src/network/source_sampler.ts

```diff
@@ -0,0 +1,216 @@
+import { ContractFunctionObj } from '@0x/base-contract';
+import { BigNumber } from '@0x/utils';
+
+import { Chain, ChainEthCallOpts } from './chain';
+import { Address, DexSample, FillData, MultiHopCallInfo, SourceSampler } from './types';
+import { ContractFunction, ContractHelper, ContractWrapperType, createContractWrapperAndHelper, GeneratedContract } from './utils';
+
+export abstract class SourceSamplerBase implements SourceSampler {
+    protected constructor() {}
+
+    public canConvertTokens(_tokenAddressPath: Address[]): boolean {
+        return true;
+    }
+
+    public abstract getSellSamplesAsync(
+        tokenAddressPath: string[],
+        takerFillAmounts: BigNumber[],
+    ): Promise<DexSample[][]>;
+
+    public abstract getBuySamplesAsync(
+        tokenAddressPath: string[],
+        makerFillAmounts: BigNumber[],
+    ): Promise<DexSample[][]>;
+
+    public abstract getMultiHopSellCallInfosAsync(
+        tokenAddressPath: Address[],
+        takerFillAmount: BigNumber,
+    ): Promise<MultiHopCallInfo[]>;
+
+    public abstract getMultiHopBuyCallInfosAsync(
+        tokenAddressPath: Address[],
+        makerFillAmount: BigNumber,
+    ): Promise<MultiHopCallInfo[]>;
+}
+
+interface OnChainSourceSamplerOptions<
+    TSellSamplerContract extends GeneratedContract,
+    TBuySamplerContract extends GeneratedContract,
+    TSellSamplerFunctionName = keyof TSellSamplerContract,
+    TBuySamplerFunctionName = keyof TBuySamplerContract
+> {
+    chain: Chain,
+    sellSamplerContractArtifactName: string;
+    buySamplerContractArtifactName: string;
+    sellSamplerContractType: ContractWrapperType<TSellSamplerContract>,
+    buySamplerContractType: ContractWrapperType<TBuySamplerContract>,
+    sellContractSellFunctionName: TSellSamplerFunctionName,
+    buyContractBuyFunctionName: TBuySamplerFunctionName,
+}
+
+export interface SamplerEthCall<TFillData, TSamplerFunctionArgs extends any[], TSamplerFunctionReturn> {
+    args: TSamplerFunctionArgs;
+    getDexSamplesFromResult(result: TSamplerFunctionReturn): DexSample<TFillData>[];
+}
+
+export type UnwrapContractFunctionReturnType<T> = T extends ContractFunctionObj<infer U> ? U : T;
+
+// Base class for a standard sampler with on-chain quote functions.
+export abstract class OnChainSourceSampler<
```

Comment:
All individual source samplers currently inherit from this class for convenience, but it isn't required. In the future we could have (e.g.) totally off-chain samplers that just inherit from the less opinionated `SourceSamplerBase` and do everything off-chain.